content/blog/elm-warrior-walkthrough/index.md

```diff
@@ -0,0 +1,818 @@
+can check our fighting spirit:
+
+```elm
+( newCoords, Warrior _ ) :: _ ->
+    if isWounded warrior then
+        ( NoGo
+        , Warrior.Wait
+        )
+
+    else
+        ( Enemy
+        , Warrior.Attack dir
+        )
+```
+
+And in `takeTurn`:
+
+```elm
+takeTurn : Warrior -> Map -> History -> Warrior.Action
+takeTurn warrior map history =
+    if not (isEnemyNearby warrior map) && isWounded warrior then
+        Warrior.Heal
+
+    else
+        allDirections
+            |> List.map (exploreDirection warrior map history)
+            |> List.sortWith
+                (\( a, _ ) ( b, _ ) -> compareExploration a b)
+            |> List.head
+            |> Maybe.map Tuple.second
+            |> Maybe.withDefault Warrior.Wait
+```
+
+What happens now?
+
+![Elm warrior pacifist](./pacifist.gif)
+
+Turns out that as soon as we move near the other warrior, we get attacked.
+Therefore, we immediately run away to heal, then try again. As much as I am
+a pacifist, we still need to find a way to pass this level.
+
+What's the solution this time? To endure a little bit of pain...
+
+Instead of running away as soon as we get slightly wounded, we can keep
+fighting for a bit and only run away when we're badly wounded.
+
+```elm
+isBadlyWounded : Warrior -> Bool
+isBadlyWounded warrior =
+    Warrior.health warrior < Warrior.maxHealth warrior // 4
+```
+
+Now, we're going to replace the use of `isWounded` inside our
+`exploreDirection` function. We still want the original function inside
+`takeTurn` because we want to heal back to full health when we're safe.
+
+```elm
+( newCoords, Warrior _ ) :: _ ->
+    if isBadlyWounded warrior then
+        ( NoGo
+        , Warrior.Wait
+        )
+
+    else
+        ( Enemy
+        , Warrior.Attack dir
+        )
+```
+
+Success!
+
+## Level X
+
+We're [here](https://ellie-app.com/9NGC6Q2YbZra1) now.
+
+![Elm warrior level ten](./level-ten.png)
+
+Ooh, items! Let's go and grab them 🏴‍☠️
+
+We'll need to add a new branch in the case statement in `exploreDirection`:
+
+```elm
+( newCoords, Item _ ) :: _ ->
+    ( CanMove (visitedCount newCoords)
+    , Warrior.Move dir
+    )
+```
+
+And change our `takeTurn` function to pickup the item if we are standing on
+it:
+
+```elm
+case
+    ( isEnemyNearby warrior map
+    , isWounded warrior
+    , Map.lookDown warrior map
+    )
+of
+    ( False, True, _ ) ->
+        Warrior.Heal
+
+    ( _, _, Item _ ) ->
+        Warrior.Pickup
+
+    _ ->
+        allDirections
+            |> List.map (exploreDirection warrior map history)
+            |> List.sortWith
+                (\( a, _ ) ( b, _ ) -> compareExploration a b)
+            |> List.head
+            |> Maybe.map Tuple.second
+            |> Maybe.withDefault Warrior.Wait
+```
+
+These changes will get us through [both levels ten and
+eleven](https://ellie-app.com/9NGZqhRjqTGa1)!
+
+## One more thing...
+
+On the last level we can notice something a bit weird:
```

Comment:
🎨 this is actually noticeable on the first fighting level, and it definitely caught my attention!
would it make sense to address it before picking up the items? not sure :-)

Reply:
Ah I couldn't be bothered to rebuild all the ellies at that point!